File: sdks/typescript/solo-client.js
export class SoloHttpError extends Error {
  constructor(message, status, body) {
    super(message);
    this.name = "SoloHttpError";
    this.status = status;
    this.body = body;
  }
}

export class SoloMcpError extends Error {
  constructor(message, error) {
    super(message);
    this.name = "SoloMcpError";
    this.error = error;
  }
}

export class SoloClient {
  #rpcId = 0;

  constructor(options = {}) {
    this.baseUrl = (options.baseUrl ?? "http://127.0.0.1:17821").replace(/\/+$/, "");
    this.bearerToken = options.bearerToken;
    this.fetchImpl = options.fetchImpl ?? globalThis.fetch;
    if (!this.fetchImpl) {
      throw new Error("SoloClient needs fetch; use Node 18+ or pass fetchImpl.");
    }
  }

  status() {
    return this.#json("GET", "/v1/status");
  }

  remember(content, options = {}) {
    const body = { content };
    if (options.sourceType !== undefined) body.source_type = options.sourceType;
    if (options.sourceId !== undefined) body.source_id = options.sourceId;
    if (options.salience !== undefined) body.salience = options.salience;
    return this.#json("POST", "/memory", body);
  }

  rememberInbox(content, options = {}) {
    return this.remember(content, {
      ...options,
      sourceType: options.sourceType ?? "solo_desktop.inbox",
    });
  }

  memoryInbox(options = {}) {
    const query = queryString({
      limit: options.limit,
    });
    return this.#json("GET", `/v1/inbox${query}`);
  }

  reviewMemory(memoryId, options = {}) {
    const body = {};
    if (options.state !== undefined) body.state = options.state;
    if (options.note !== undefined) body.note = options.note;
    return this.#json(
      "POST",
      `/v1/inbox/${encodeURIComponent(memoryId)}/review`,
      body,
    );
  }

  recall(query, options = {}) {
    const body = { query };
    if (options.limit !== undefined) body.limit = options.limit;
    return this.#json("POST", "/memory/search", body);
  }

  context(query, options = {}) {
    const body = { query };
    if (options.subject !== undefined) body.subject = options.subject;
    if (options.windowDays !== undefined) body.window_days = options.windowDays;
    if (options.limit !== undefined) body.limit = options.limit;
    return this.#json("POST", "/memory/context", body);
  }

  factsAbout(subject, options = {}) {
    const query = queryString({
      subject,
      predicate: options.predicate,
      since_ms: options.sinceMs,
      until_ms: options.untilMs,
      include_as_object: options.includeAsObject,
      limit: options.limit,
    });
    return this.#json("GET", `/memory/facts_about${query}`);
  }

  entities(query, options = {}) {
    const params = queryString({ query, limit: options.limit });
    return this.#json("GET", `/memory/entities${params}`);
  }

  inspect(memoryId) {
    return this.#json(
      "GET",
      `/memory/${encodeURIComponent(memoryId)}`,
      undefined,
    );
  }

  update(memoryId, content) {
    return this.#json(
      "PATCH",
      `/memory/${encodeURIComponent(memoryId)}`,
      { content },
    );
  }

  forget(memoryId, options = {}) {
    const reason = options.reason ? `?reason=${encodeURIComponent(options.reason)}` : "";
    return this.#json(
      "DELETE",
      `/memory/${encodeURIComponent(memoryId)}${reason}`,
      undefined,
    );
  }

  listDocuments(options = {}) {
    const query = queryString({
      limit: options.limit,
      offset: options.offset,
      include_forgotten: options.includeForgotten,
    });
    return this.#json("GET", `/memory/documents${query}`);
  }

  ingestDocument(path) {
    return this.#json("POST", "/memory/documents", { path });
  }

  searchDocuments(query, options = {}) {
    const body = { query };
    if (options.limit !== undefined) body.limit = options.limit;
    return this.#json("POST", "/memory/documents/search", body);
  }

  inspectDocument(docId) {
    return this.#json(
      "GET",
      `/memory/documents/${encodeURIComponent(docId)}`,
      undefined,
    );
  }

  forgetDocument(docId) {
    return this.#json(
      "DELETE",
      `/memory/documents/${encodeURIComponent(docId)}`,
      undefined,
    );
  }

  recentMemories(options = {}) {
    const query = queryString({
      kind: "episode",
      limit: options.limit ?? 100,
      cursor: options.cursor,
      since_ms: options.sinceMs,
      until_ms: options.untilMs,
    });
    return this.#json("GET", `/v1/graph/nodes${query}`);
  }

  async mcpConnect(clientInfo = { name: "solo-sdk", version: "0.0.0" }) {
    const initialized = await this.mcpInitialize(clientInfo);
    if (!initialized.sessionId) {
      throw new Error("Solo MCP initialize did not return an Mcp-Session-Id header.");
    }
    await this.mcpNotifyInitialized(initialized);
    return initialized;
  }

  async mcpInitialize(clientInfo = { name: "solo-sdk", version: "0.0.0" }) {
    const response = await this.#request("POST", "/mcp", {
      jsonrpc: "2.0",
      id: this.#nextRpcId(),
      method: "initialize",
      params: {
        protocolVersion: "2025-03-26",
        capabilities: {},
        clientInfo,
      },
    });
    const raw = await this.#parseJson(response);
    const result = this.#jsonRpcResult(raw);
    return {
      sessionId: response.headers.get("mcp-session-id"),
      result,
      raw,
    };
  }

  async mcpNotifyInitialized(session) {
    const sessionId = this.#sessionId(session);
    const response = await this.#request(
      "POST",
      "/mcp",
      { jsonrpc: "2.0", method: "notifications/initialized", params: {} },
      { "Mcp-Session-Id": sessionId },
    );
    await response.text();
  }

  async mcpListTools(session) {
    const sessionId = this.#sessionId(session);
    const raw = await this.#json("POST", "/mcp", {
      jsonrpc: "2.0",
      id: this.#nextRpcId(),
      method: "tools/list",
    }, { "Mcp-Session-Id": sessionId });
    const result = this.#jsonRpcResult(raw);
    return result?.tools ?? [];
  }

  async mcpCallTool(session, name, args = {}) {
    const sessionId = this.#sessionId(session);
    const raw = await this.#json("POST", "/mcp", {
      jsonrpc: "2.0",
      id: this.#nextRpcId(),
      method: "tools/call",
      params: {
        name,
        arguments: args ?? {},
      },
    }, { "Mcp-Session-Id": sessionId });
    return this.#jsonRpcResult(raw);
  }

  async #json(method, path, body, headers = {}) {
    const response = await this.#request(method, path, body, headers);
    return this.#parseJson(response);
  }

  async #request(method, path, body, extraHeaders = {}) {
    const headers = {
      Accept: "application/json",
      ...extraHeaders,
    };
    if (body !== undefined) headers["Content-Type"] = "application/json";
    if (this.bearerToken) headers.Authorization = `Bearer ${this.bearerToken}`;

    let response;
    try {
      response = await this.fetchImpl(`${this.baseUrl}${path}`, {
        method,
        headers,
        body: body === undefined ? undefined : JSON.stringify(body),
      });
    } catch (error) {
      const detail = error instanceof Error ? error.message : String(error);
      throw new Error(
        `Could not reach Solo daemon at ${this.baseUrl}. Start Solo Desktop/tray or run solo daemon. ${detail}`,
      );
    }

    if (!response.ok) {
      const text = await response.text();
      const parsed = tryJson(text);
      const message = parsed?.error ?? (text || `Solo HTTP ${response.status}`);
      throw new SoloHttpError(message, response.status, parsed ?? text);
    }
    return response;
  }

  async #parseJson(response) {
    const text = await response.text();
    return text ? JSON.parse(text) : {};
  }

  #nextRpcId() {
    this.#rpcId += 1;
    return this.#rpcId;
  }

  #sessionId(session) {
    if (typeof session === "string") {
      if (!session) throw new Error("Solo MCP session id is required.");
      return session;
    }
    if (!session?.sessionId) throw new Error("Solo MCP session id is required.");
    return session.sessionId;
  }

  #jsonRpcResult(raw) {
    if (raw?.error) {
      const message = raw.error.message ?? raw.error.error ?? "Solo MCP JSON-RPC error";
      throw new SoloMcpError(message, raw.error);
    }
    return raw?.result;
  }
}

function tryJson(text) {
  try {
    return text ? JSON.parse(text) : undefined;
  } catch {
    return undefined;
  }
}

function queryString(params) {
  const query = new URLSearchParams();
  for (const [key, value] of Object.entries(params)) {
    if (value !== undefined && value !== null) query.set(key, String(value));
  }
  const text = query.toString();
  return text ? `?${text}` : "";
}

File: sdks/python/solo_client.py
import json
from urllib.error import HTTPError, URLError
from urllib.parse import quote, urlencode
from urllib.request import Request, urlopen


class SoloHttpError(RuntimeError):
    def __init__(self, message, status, body):
        super().__init__(message)
        self.status = status
        self.body = body


class SoloMcpError(RuntimeError):
    def __init__(self, message, error):
        super().__init__(message)
        self.error = error


class SoloClient:
    def __init__(
        self,
        base_url="http://127.0.0.1:17821",
        bearer_token=None,
        timeout=10.0,
    ):
        self.base_url = base_url.rstrip("/")
        self.bearer_token = bearer_token
        self.timeout = timeout
        self._rpc_id = 0

    def status(self):
        return self._json("GET", "/v1/status")

    def remember(
        self,
        content,
        source_type=None,
        source_id=None,
        salience=None,
    ):
        body = {"content": content}
        if source_type is not None:
            body["source_type"] = source_type
        if source_id is not None:
            body["source_id"] = source_id
        if salience is not None:
            body["salience"] = salience
        return self._json("POST", "/memory", body)

    def remember_inbox(
        self,
        content,
        source_type="solo_desktop.inbox",
        source_id=None,
        salience=None,
    ):
        return self.remember(
            content,
            source_type=source_type,
            source_id=source_id,
            salience=salience,
        )

    def memory_inbox(self, limit=None):
        params = {}
        if limit is not None:
            params["limit"] = limit
        return self._json("GET", f"/v1/inbox{_query(params)}")

    def review_memory(self, memory_id, state=None, note=None):
        body = {}
        if state is not None:
            body["state"] = state
        if note is not None:
            body["note"] = note
        return self._json(
            "POST",
            f"/v1/inbox/{_quote(memory_id)}/review",
            body,
        )

    def recall(self, query, limit=None):
        body = {"query": query}
        if limit is not None:
            body["limit"] = limit
        return self._json("POST", "/memory/search", body)

    def context(self, query, subject=None, window_days=None, limit=None):
        body = {"query": query}
        if subject is not None:
            body["subject"] = subject
        if window_days is not None:
            body["window_days"] = window_days
        if limit is not None:
            body["limit"] = limit
        return self._json("POST", "/memory/context", body)

    def facts_about(
        self,
        subject,
        predicate=None,
        since_ms=None,
        until_ms=None,
        include_as_object=None,
        limit=None,
    ):
        params = {"subject": subject}
        if predicate is not None:
            params["predicate"] = predicate
        if since_ms is not None:
            params["since_ms"] = since_ms
        if until_ms is not None:
            params["until_ms"] = until_ms
        if include_as_object is not None:
            params["include_as_object"] = _bool_query(include_as_object)
        if limit is not None:
            params["limit"] = limit
        return self._json("GET", f"/memory/facts_about{_query(params)}")

    def entities(self, query, limit=None):
        params = {"query": query}
        if limit is not None:
            params["limit"] = limit
        return self._json("GET", f"/memory/entities{_query(params)}")

    def inspect(self, memory_id):
        return self._json("GET", f"/memory/{_quote(memory_id)}")

    def update(self, memory_id, content):
        return self._json(
            "PATCH",
            f"/memory/{_quote(memory_id)}",
            {"content": content},
        )

    def forget(self, memory_id, reason=None):
        suffix = f"?reason={_quote(reason)}" if reason else ""
        return self._json("DELETE", f"/memory/{_quote(memory_id)}{suffix}")

    def list_documents(self, limit=None, offset=None, include_forgotten=None):
        params = {}
        if limit is not None:
            params["limit"] = limit
        if offset is not None:
            params["offset"] = offset
        if include_forgotten is not None:
            params["include_forgotten"] = _bool_query(include_forgotten)
        return self._json("GET", f"/memory/documents{_query(params)}")

    def ingest_document(self, path):
        return self._json("POST", "/memory/documents", {"path": path})

    def search_documents(self, query, limit=None):
        body = {"query": query}
        if limit is not None:
            body["limit"] = limit
        return self._json("POST", "/memory/documents/search", body)

    def inspect_document(self, doc_id):
        return self._json("GET", f"/memory/documents/{_quote(doc_id)}")

    def forget_document(self, doc_id):
        return self._json("DELETE", f"/memory/documents/{_quote(doc_id)}")

    def recent_memories(
        self,
        limit=100,
        cursor=None,
        since_ms=None,
        until_ms=None,
    ):
        params = {"kind": "episode"}
        if limit is not None:
            params["limit"] = limit
        if cursor is not None:
            params["cursor"] = cursor
        if since_ms is not None:
            params["since_ms"] = since_ms
        if until_ms is not None:
            params["until_ms"] = until_ms
        return self._json("GET", f"/v1/graph/nodes{_query(params)}")

    def mcp_connect(self, client_name="solo-sdk", client_version="0.0.0"):
        session = self.mcp_initialize(client_name, client_version)
        if not session["session_id"]:
            raise RuntimeError("Solo MCP initialize did not return an Mcp-Session-Id header.")
        self.mcp_notify_initialized(session)
        return session

    def mcp_initialize(self, client_name="solo-sdk", client_version="0.0.0"):
        payload, headers = self._send("POST", "/mcp", {
            "jsonrpc": "2.0",
            "id": self._next_rpc_id(),
            "method": "initialize",
            "params": {
                "protocolVersion": "2025-03-26",
                "capabilities": {},
                "clientInfo": {"name": client_name, "version": client_version},
            },
        })
        result = self._json_rpc_result(payload)
        return {
            "session_id": headers.get("Mcp-Session-Id"),
            "result": result,
            "raw": payload,
        }

    def mcp_notify_initialized(self, session):
        session_id = self._session_id(session)
        self._send(
            "POST",
            "/mcp",
            {"jsonrpc": "2.0", "method": "notifications/initialized", "params": {}},
            {"Mcp-Session-Id": session_id},
        )

    def mcp_list_tools(self, session):
        session_id = self._session_id(session)
        payload = self._json(
            "POST",
            "/mcp",
            {"jsonrpc": "2.0", "id": self._next_rpc_id(), "method": "tools/list"},
            {"Mcp-Session-Id": session_id},
        )
        return self._json_rpc_result(payload).get("tools", [])

    def mcp_call_tool(self, session, name, arguments=None):
        session_id = self._session_id(session)
        payload = self._json(
            "POST",
            "/mcp",
            {
                "jsonrpc": "2.0",
                "id": self._next_rpc_id(),
                "method": "tools/call",
                "params": {
                    "name": name,
                    "arguments": {} if arguments is None else arguments,
                },
            },
            {"Mcp-Session-Id": session_id},
        )
        return self._json_rpc_result(payload)

    def _json(self, method, path, body=None, headers=None):
        payload, _headers = self._send(method, path, body, headers)
        return payload

    def _send(self, method, path, body=None, headers=None):
        data = json.dumps(body).encode("utf-8") if body is not None else None
        req = Request(f"{self.base_url}{path}", data=data, method=method)
        req.add_header("Accept", "application/json")
        if body is not None:
            req.add_header("Content-Type", "application/json")
        if self.bearer_token:
            req.add_header("Authorization", f"Bearer {self.bearer_token}")
        for name, value in (headers or {}).items():
            req.add_header(name, value)

        try:
            with urlopen(req, timeout=self.timeout) as response:
                raw = response.read().decode("utf-8")
                return (json.loads(raw) if raw else {}, response.headers)
        except HTTPError as exc:
            raw = exc.read().decode("utf-8", errors="replace")
            body = _try_json(raw)
            message = body.get("error") if isinstance(body, dict) else raw
            raise SoloHttpError(message or f"Solo HTTP {exc.code}", exc.code, body or raw) from exc
        except URLError as exc:
            raise RuntimeError(
                f"Could not reach Solo daemon at {self.base_url}. "
                "Start Solo Desktop/tray or run solo daemon."
            ) from exc

    def _next_rpc_id(self):
        self._rpc_id += 1
        return self._rpc_id

    def _session_id(self, session):
        if isinstance(session, str):
            if not session:
                raise RuntimeError("Solo MCP session id is required.")
            return session
        session_id = session.get("session_id")
        if not session_id:
            raise RuntimeError("Solo MCP session id is required.")
        return session_id

    def _json_rpc_result(self, payload):
        error = payload.get("error")
        if error:
            message = (
                error.get("message") if isinstance(error, dict) else "Solo MCP JSON-RPC error"
            )
            raise SoloMcpError(message or "Solo MCP JSON-RPC error", error)
        return payload.get("result", {})


def _try_json(text):
    try:
        return json.loads(text) if text else None
    except json.JSONDecodeError:
        return None


def _quote(value):
    return quote(str(value), safe="")


def _query(params):
    if not params:
        return ""
    return "?" + urlencode({key: str(value) for key, value in params.items()})


def _bool_query(value):
    return "true" if value else "false"
